NewFeature: Implement Model for Building-Integrated Photovoltaic Thermal (BIPVT)

khaddad opened the pull request:
### Pull request overview

Currently EnergyPlus supports modelling PV panels mounted on building exterior surfaces. The thermal performance of the panels is modelled using a solar collector with a constant thermal conversion efficiency to predict the temperature of the PV layer. This new feature expands the capability of the current solar collector model for PV panels to account for the variation of the thermal conversion efficiency with the flow parameters inside the air channel behind the panels and ambient conditions (BIPVT).

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

khaddad committed:
design document for advanced BIPVT model

khaddad committed:
1-update idd file to include BIPVT object parameters
2-add structures for new BIPVT objects
3-process input data of BIPVT objects
4-add model to calculate thermal performance of BIPVT objects

sebastienbrideau committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into HEAD of bipvt

sebastienbrideau committed:
fix bug

khaddad committed:
merge e+ developerment branch

khaddad committed:
Update BIPVT model implementation to account for mixer of bypass air and air from BIPVT channel.

khaddad committed:
Update other side boundary condition for BIPVT model

Add link between BIPVT model and PV electric model

sebastienbrideau committed:
Changed method for calculating absorbed radiation on BIPVT wall. taoalpha method is based on method for PV from Duffie and Beckman. This method gives acceptable results for mate surfaces as well (wall) and for PV backing, so I am using that for those as well.

Some values are hard coded. That needs to be fixed.

Compiles and runs.

sebastienbrideau committed:
Merge branch 'bipvt' of https://github.com/khaddad/EnergyPlus into bipvt

sebastienbrideau committed:
This commit:
-adds an operation mode for stagnation (zero flow).
-adds depth_channel (hard coded for now) required for calculating internal heat transfer coefficient
-adds hard coded (for now) external heat transfer coefficient, and related variables
-adds calculations of air thermal properties required for internal heat transfer coefficient
-changed equivalent surrounding temperature (for radiation exchange calculation) tsurr

sebastienbrideau committed:
added minor comments

sebastienbrideau committed:
changed minor comment

khaddad committed:
merge with e+ develop

khaddad committed:
resolve conflicts with e+ merge

khaddad committed:
merge with main branch

khaddad committed:
set tsuurK and tpvgK at the beginning of the iteration in the subroutine

khaddad committed:
merge with develop branch

khaddad committed:
expand idd for bipvt model

khaddad committed:
merge with develop branch

khaddad committed:
correct syntax of bipvt object

khaddad committed:
correct code to access NumPVT parameter

khaddad committed:
update bipvt method to use new added parameters; initialize new parameters in header file

khaddad committed:
update file format

khaddad committed:
apply format to files

khaddad committed:
apply format to file

khaddad committed:
update file format

sebastienbrideau committed:
Fixes minor mistake with assignment of inputs to correct variable name for depth_channel

sebastienbrideau committed:
Merge branch 'bipvt' of https://github.com/khaddad/EnergyPlus into bipvt

sebastienbrideau committed:
fix clang format error caused by my last commit (hopefully)

sebastienbrideau committed:
This commit fixes a few bugs, following some initial testing
-change to prandtl number equation
-change to input for PV module backing material tao_alpha product
-change to collector temp in GetPVTTsColl()
-change to incident solar radiation "g" to account for incident angle modifier for beam, sky diffuse, and ground diffuse
-change to definition of t2

khaddad committed:
merge EnergyPlus develop branch

khaddad committed:
merge develop branch

khaddad committed:
modify expression for outside building surface temp

khaddad committed:
remove duplicate bipvt entry from idd file

khaddad committed:
additional changes to bipvt model

khaddad committed:
update file format

khaddad committed:
add check to limit min value of tfavg

sebastienbrideau committed:
change name of one parameter

khaddad committed:
Merge remote-tracking branch 'upstream/develop' into develop

khaddad committed:
Merge branch 'bipvt' of https://github.com/khaddad/EnergyPlus into bipvt

khaddad committed:
Update variable used for outside surface temp.

khaddad committed:
Add test case for BIPVT model

khaddad committed:
remove extra tab

khaddad committed:
Use correct variable for outdoor dry-bulb temp. to calculate collector temp. (Simple PVT).

khaddad committed:
Add test file "ShopWithBIPVT" to cmake list file

khaddad committed:
Update test case for BIPVT to use PV modules performance type of
"EquivalentOne-Diode"

Use one OtherSideConditionsModel for each surface

sebastienbrideau committed:
Changes to air properties calculations
Changes to heat transfer coefficient calculations (external)
Changes to variable declarations (mostly for clarity)

khaddad committed:
Merge remote-tracking branch 'upstream/develop' into develop

khaddad committed:
Fix variable used for outside air temp of solar collector

khaddad committed:
Merge branch 'develop' into bipvt

khaddad committed:
Update format

khaddad committed:
Update 'const' with 'constexpr'

khaddad committed:
Update file format

khaddad committed:
Remove extra tabs and spaces

khaddad committed:
Merge remote-tracking branch 'upstream/develop' into develop

khaddad committed:
Update BIPVT model test case

nrel-bot-2c commented:
@khaddad @myoldmopar it has been 28 days since this pull request was last updated.

khaddad committed:
Merge remote-tracking branch 'upstream/develop' into develop

khaddad committed:
Update BIPVT model test case

sebastienbrideau committed:
-adds input-output docs
-adds engineering reference docs
-changes to forced internal heat transfer coefficients calcs in PhotovoltaicThermalCollectors.cc

khaddad committed:
Remove unused BIPVT model variables

Update BIPVT model test case

Update IDD file

khaddad committed:
Run clang format

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -71297,6 +71297,96 @@ SolarCollectorPerformance:PhotovoltaicThermal:Simple,
+SolarCollectorPerformance:PhotovoltaicThermal:BIPVT,
+        \memo Thermal performance parameters for Building-Integrated Photovoltaic-Thermal (BIPVT) solar collector.
+  A1 ,  \field Name
+        \reference FlatPlateBIPVTParameters
```
If I am following correctly, this should be `FlatPlatePVTParameters` so that it matches the object-list name here:
```
SolarCollector:FlatPlate:PhotovoltaicThermal,
  A3 , \field Photovoltaic-Thermal Model Performance Name
       \type object-list
       \object-list FlatPlatePVTParameters
```

mjwitte replied:
And Name should be tagged as `\required-field`, same for `SolarCollectorPerformance:PhotovoltaicThermal:Simple` and `SolarCollector:FlatPlate:PhotovoltaicThermal`.

khaddad replied:
This field is updated to 'FlatPlatePVTParameters'

mjwitte commented on c:
```diff
@@ -0,0 +1,515 @@
+[33mcommit 2b1fbaf3a1dc9fae0c19d08cde4f1d343a083eb8[m[33m ([m[1;36mHEAD -> [m[1;32mdevelop[m[33m, [m[1;31morigin/develop[m[33m, [m[1;31morigin/HEAD[m[33m)[m
```
This file shouldn't be in the branch.

khaddad replied:
removed from the repo

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -71297,6 +71297,96 @@ SolarCollectorPerformance:PhotovoltaicThermal:Simple,
+        \required-field
+        \type real
+        \units m
+        \minimum> 0.0
```
I believe other solar/PV models in EnergyPlus get the geometry from the attached surface. Is there a need for these inputs here? Is the area of the surface being used? If so, then what happens if these inputs are not consistent with the surface area? Perhaps these would be better as a ratio to the surface dimensions?

sebastienbrideau replied:
Commit c85bc0105bbe9da675950f197498f7d4a7320acc

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -71297,6 +71297,96 @@ SolarCollectorPerformance:PhotovoltaicThermal:Simple,
+        \default 0.85
+        \minimum> 0.0
+        \maximum< 1.00
+  N8 ,  \field Fraction of PV cell area to PV module area
```
Title case. "Fraction of PV Cell Area to PV Modul Area". Check other field names.

khaddad replied:
changed as suggested

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -71297,6 +71297,96 @@ SolarCollectorPerformance:PhotovoltaicThermal:Simple,
+        \default 0.9
+        \minimum> 0.0
+        \maximum< 1.00
+  N9 ,  \field PV Module Thermal Resistance - Top
```
How about "PV Module Top Thermal Resistance" and "PV Module Bottom Thermal Resistance"?

sebastienbrideau replied:
commit 3f3e6f2c90d651f8c4d10efae4619e89d3c7fd80

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -71297,6 +71297,96 @@ SolarCollectorPerformance:PhotovoltaicThermal:Simple,
+        \default 0.85
+        \minimum> 0.0
+        \maximum< 1.00
+  N12 , \field Back of Modules Material Longwave Emissivity
```
For this pair, how about "PV Module Front Longwave Emissivity" and "PV Module Back Longwave Emissivity"?

sebastienbrideau replied:
commit 18a19110d5f42d01c2fe303135e5c954e3d59c94

mjwitte commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -104,10 +107,14 @@ namespace PhotovoltaicThermalCollectors {
     //  Simple PV/T model just converts incoming solar to electricity and temperature rise of a working fluid.
 
     int constexpr SimplePVTmodel(1001);
+    int constexpr BIPVTmodel(1002);
```
This should be an enum class, starting with `Invalid` and ending with `Num`, see [example](https://github.com/NREL/EnergyPlus/blob/4a85292a9c2c480dea9ffb11c03b8c63b16c9135/src/EnergyPlus/CondenserLoopTowers.hh#L72-L81).

khaddad replied:
An enum class is now defined in the header file instead.

mjwitte commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -147,78 +154,68 @@ namespace PhotovoltaicThermalCollectors {
         //       DATE WRITTEN   June 2008
-        //       MODIFIED       na
+        //       MODIFIED       K. Haddad, March 2020, add support to read inputs for BIPVT objects
         //       RE-ENGINEERED  na
```
We no longer use these types of comments, these two lines can be deleted.

khaddad replied:
line 160 deleted

mjwitte commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -147,78 +154,68 @@ namespace PhotovoltaicThermalCollectors {
         int NumNumbers;          // Number of Numbers for each GetObjectItem call
         int IOStatus;            // Used in GetObjectItem
         bool ErrorsFound(false); // Set to true if errors in input, fatal at end of routine
+        using ScheduleManager::GetScheduleIndex;
```
Current style is to avoid `using` statements.

khaddad replied:
It turns out this statement is not needed and is then removed from the method.

mjwitte commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -147,78 +154,68 @@ namespace PhotovoltaicThermalCollectors {
+
+    void GetPVTSimpleCollectorsInput(EnergyPlusData &state, int NumSimplePVTPerform, Array1D<SimplePVTModelStruct> &tmpSimplePVTperf)
+    {
+        // SUBROUTINE INFORMATION:
+        //       AUTHOR         B. Griffith
+        //       DATE WRITTEN   June 2008
+        //       MODIFIED       K. Haddad, March 2020, subroutine created from original code in subroutine
+        //                      "GetPVTcollectorsInput" to read inputs for PVT objects
+        //       RE-ENGINEERED  na
```
Boilerplate comments like this are no longer used.

khaddad replied:
lines 200-205 are removed
Also removed these comments from all the subroutines that we added. We left any existing similar comments in the subroutines that are currently part of EnergyPlus.

mjwitte commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -181,13 +225,36 @@ namespace PhotovoltaicThermalCollectors {
 
         void calculate(EnergyPlusData &state);
 
+        void SimplePVTcalculate(EnergyPlusData &state);
+
+        void BIPVTcalculate(EnergyPlusData &state);
+
+        void BIPVT_MaxHeatGain_calculate(
+            EnergyPlusData &state, Real64 tsp, std::string Mode, Real64 &bfr, Real64 &q, Real64 &tmixed, Real64 &ThEff, Real64 &tpv);
+
+        void solve_lin_sys_back_sub(Real64 jj[9], Real64 f[3], Real64 (&y)[3]);
```
If you look in the wiki and the code, you'll find conflicting guidance on naming conventions (sorry). The current preferred style is camelCase for function names, starting with the active word. So, these would be:
`calculateSimplePVT, calculateBIPVT, calculateBIPVTMaxHeatGain, solveLinSysBackSub`.

khaddad replied:
changed names of methods as suggested in this comment

khaddad replied:
names of methods updated as suggested

mjwitte reviewed (commented):
@khaddad Some initial review comments.
There are [build warnings](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-4dd29894665d4d7d1a45f4ccc980909fe3fde87f-PR9594-x86_64-MacOS-10.17-clang-13.0.0.html) to clean up.

khaddad committed:
This commit addresses comments from Energyplus team.
1) Delete extra file c
2) Replace use of 'constexp' with enum class in the header for PVT
3) Update names of methods
4) Remove comments at the beginning of methods
5) Modify variable name in idd file for BIPVT

khaddad committed:
Update enum class variable name to begin with capital letter

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into bipvt

mjwitte committed:
Doc and IDD cleanup

mjwitte reviewed (commented):
@khaddad Thanks for addressing some of the comments. 
I pushed up some minor doc and IDD cleanup (easier to just do the edits than to mark them).
There are still some build warnings to fix.

mjwitte commented:
Also, when I build the engineering reference locally, there are some errors.
```
1>engineering-reference Pass 3: TOC OK, Number of entries in TOC = 63
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1325::Extra \right.l.1325 ...au\alpha (0)_{clad}\cdot G_{clad}\right) \cdot \left(1-F_{module}\r... I'm ignoring a \right that had no matching \left.]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1385::Missing } inserted.<inserted text> } l.1385 \(T_{PV\) is the temperature of the PV cells (\(^{\circ}C\)) I've inserted something that you may have forgotten. (See the <inserted text> above.) With luck, this will get me unwedged. But if you really didn't forget anything, try typing `2' now; then my insertion and my current dilemma will both disappear.]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1643::Paragraph ended before \begin was complete.<to be read again> \par l.1643]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1643::Missing $ inserted.<inserted text> $ l.1643]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1643::Display math should end with $$.<to be read again> \tex_par:D l.1643]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1673::Missing delimiter (. inserted).<to be read again> d l.1673 ...eft|T_{2}-T_{1}|\right d^{3}}{\nu\alpha}]
1>[LATEXERROR::doc\engineering-reference\src\special-modules-reporting\resilience-metrics.tex:None::LaTeX Error: \begin{equation} on input line 1639 ended by \end{document}.]
```

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -77,6 +77,8 @@ namespace PhotovoltaicThermalCollectors {
         FIXED
     };
 
+    extern int NumPVT; // number of transpired collectors in model
```
Not sure why I am only getting to this PR now but ... here goes. No free-floating global variables like this. This should be part of `state`.

khaddad replied:
updated

Myoldmopar replied:
Did your change get pushed?  It's still showing up here in the latest Git diff.  We can't have non-const globals, and this extern indicates it is still here.

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -77,6 +77,8 @@ namespace PhotovoltaicThermalCollectors {
         FIXED
```
`enum` that will be read from the IDF need to start with `Invalid = -1` and end with `Num`.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -97,6 +99,46 @@ namespace PhotovoltaicThermalCollectors {
+        Real64 HrPlen;                // Modeled radiation coef for OSCM (W/m2-C)
+        Real64 HcPlen;                // Modeled Convection coef for OSCM (W/m2-C)
+
+        // Default Constructor
```
Initialize in declaration, remove default construct.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -118,21 +160,30 @@ namespace PhotovoltaicThermalCollectors {
         }
     };
 
+    enum class PVTModelTypes
```
`Type`, not `Types`.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -118,21 +160,30 @@ namespace PhotovoltaicThermalCollectors {
+    enum class PVTModelTypes
+    {
+        Invalid = -1,
+        SimplePVTmodel = 1001,
```
No need to use the word `PVTmodel` since it is already in the `Type` name and will have to be used as `PVTModelType::SimplePVTmodel`. `Simple` is sufficient. Do these absolutely have to be 1001 and 1002 for some reason?

khaddad replied:
updated. not sure yet about the 1001. It was already in the code for 'Simple' model and I used 1002 for 'BIPVT'

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -150,14 +201,15 @@ namespace PhotovoltaicThermalCollectors {
-              HVACOutletNodeNum(0), DesignVolFlowRate(0.0), DesignVolFlowRateWasAutoSized(false), MaxMassFlowRate(0.0), MassFlowRate(0.0),
-              AreaCol(0.0), BypassDamperOff(true), CoolingUseful(false), HeatingUseful(false), MySetPointCheckFlag(true), MyOneTimeFlag(true),
-              SetLoopIndexFlag(true)
+            : Type(DataPlant::PlantEquipmentType::Invalid), WPlantLoc{}, EnvrnInit(true), SizingInit(true), SurfNum(0), PVnum(0), PVfound(false),
```
Initialize in declaration.

khaddad replied:
Most of these variables were already there? I can help with existing code changes if needed.

Myoldmopar replied:
Yeah, most of them were already there.  We are trying to move away from this type of initializer list constructor because it makes for weird diffs and is more difficult to see how a variable is initialized.  You don't have to go change everything already there, but if you are adding new members to a struct, we prefer that you at least initialize those new variables right where they are declared on the class.  This isn't an absolute show stopper but if you have any more changes to make, it would be nice to add this to them.

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -181,13 +233,36 @@ namespace PhotovoltaicThermalCollectors {
+
+        void calculateBIPVT(EnergyPlusData &state);
+
+        void calculateBIPVTMaxHeatGain(
```
Strings should be passed by `const &` not by value. Even better, pass a `string_view`. And better still, pass an enum if that is how you are using the string as the name suggests.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -181,13 +233,36 @@ namespace PhotovoltaicThermalCollectors {
+        void calculateBIPVTMaxHeatGain(
+            EnergyPlusData &state, Real64 tsp, std::string Mode, Real64 &bfr, Real64 &q, Real64 &tmixed, Real64 &ThEff, Real64 &tpv);
+
+        void solveLinSysBackSub(Real64 jj[9], Real64 f[3], Real64 (&y)[3]);
```
Use C++ parameters here. Also, we already have `LUdecomposition`/`LUsolution` and `LUDCMP`/`LUBKSB`.

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -181,13 +233,36 @@ namespace PhotovoltaicThermalCollectors {
+    void GetMainPVTInput(EnergyPlusData &state,
+                         int NumPVT,
+                         Array1D<PVTCollectorStruct> &PVT,
+                         Array1D<SimplePVTModelStruct> tmpSimplePVTperf,
```
Pass arrays by `const &` not value.

khaddad replied:
@amir, the array is being allocated in the method and this is why it was passed by reference?

Myoldmopar replied:
But `tmpSimplePVTperf` is being passed by value, which will copy the array.  Just pass it by const &.

Also, since we are already passing state everywhere, you are welcome to just put them on state itself like we do everywhere else in EnergyPlus.  Presumably they already live there anyway.

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -198,6 +273,12 @@ namespace PhotovoltaicThermalCollectors {
 
     int GetAirOutletNodeNum(EnergyPlusData &state, std::string_view PVTName, bool &ErrorsFound);
 
+    void GetPVTmodelIndex(EnergyPlusData &state, int const SurfacePtr, int &PVTIndex);
```
Return the index through the return value, not a parameter.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.hh:
```diff
@@ -198,6 +273,12 @@ namespace PhotovoltaicThermalCollectors {
+
+    void SetPVTQdotSource(EnergyPlusData &state, int const PVTNum, Real64 const QSource);
+
+    void GetPVTTsColl(EnergyPlusData &state, int const PVTNum, Real64 &TsColl);
```
Return `TsCoil` through the return value.

khaddad replied:
we kept this as is. We followed the same approach as for 'CellIntegration::TranspiredCollector'.

amirroth commented on src/EnergyPlus/Photovoltaics.cc:
```diff
@@ -427,6 +428,14 @@ namespace Photovoltaics {
                                               " are using exterior vented surface = " + state.dataPhotovoltaic->PVarray(PVnum).SurfaceName);
                         ErrorsFound = true;
+                    } else if (state.dataPhotovoltaic->PVarray(dupPtr).CellIntegrationMode == CellIntegration::PVTSolarCollector) {
+                        ShowSevereError(state, cCurrentModuleObject + ": problem detected with multiple PV arrays.");
```
`format`

khaddad replied:
clang format applied

khaddad replied:
now used a format approach in the error message

amirroth commented on src/EnergyPlus/Photovoltaics.cc:
```diff
@@ -427,6 +428,14 @@ namespace Photovoltaics {
                                               state.dataPhotovoltaic->PVarray(dupPtr).Name +
                                               " are using exterior vented surface = " + state.dataPhotovoltaic->PVarray(PVnum).SurfaceName);
                         ErrorsFound = true;
+                    } else if (state.dataPhotovoltaic->PVarray(dupPtr).CellIntegrationMode == CellIntegration::PVTSolarCollector) {
```
May want to use `auto &thisPVarray = state.dataPhotovoltaic->PVarray(dupPtr);` shortcut in this code.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/Photovoltaics.cc:
```diff
@@ -843,6 +852,7 @@ namespace Photovoltaics {
         // collect statements that assign to variables tied to output variables
 
         // Using/Aliasing
+        using PhotovoltaicThermalCollectors::SetPVTQdotSource;
```
We are not using `using` statements anymore.

khaddad replied:
This was updated to use 'auto' variable setting approach.

amirroth commented on src/EnergyPlus/Photovoltaics.cc:
```diff
@@ -995,7 +1007,13 @@ namespace Photovoltaics {
             } break;
             case CellIntegration::PVTSolarCollector: {
-                // add calls to PVT models here
+                GetPVTTsColl(state, state.dataPhotovoltaic->PVarray(PVnum).PVTPtr, state.dataPhotovoltaic->PVarray(PVnum).SNLPVCalc.Tback);
```
`auto &thisPVarray` would make this more readable.

khaddad replied:
this was updated in our code. Note though that existing code, not ours, does refer to variables using "state."

amirroth replied:
I completely understand. But as there is no reasonable way of telling people "go clean up your old code", what I am doing instead is opportunistically pointing out cleanup opportunities as people are doing normal bug fixes and updates. It's not a "you broke it, you fix it". It's more like "you're changing it now, you may as well also fix it."

sebastienbrideau committed:
Fixes bugs in doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/solar-collectors.tex

sebastienbrideau commented:
> 1>engineering-reference Pass 3: TOC OK, Number of entries in TOC = 63
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1325::Extra \right.l.1325 ...au\alpha (0)_{clad}\cdot G_{clad}\right) \cdot \left(1-F_{module}\r... I'm ignoring a \right that had no matching \left.]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1385::Missing } inserted.<inserted text> } l.1385 \(T_{PV\) is the temperature of the PV cells (\(^{\circ}C\)) I've inserted something that you may have forgotten. (See the <inserted text> above.) With luck, this will get me unwedged. But if you really didn't forget anything, try typing `2' now; then my insertion and my current dilemma will both disappear.]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1643::Paragraph ended before \begin was complete.<to be read again> \par l.1643]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1643::Missing $ inserted.<inserted text> $ l.1643]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1643::Display math should end with $$.<to be read again> \tex_par:D l.1643]
1>[LATEXERROR::doc\engineering-reference\src\simulation-models-encyclopedic-reference-004\solar-collectors.tex:1673::Missing delimiter (. inserted).<to be read again> d l.1673 ...eft|T_{2}-T_{1}|\right d^{3}}{\nu\alpha}]
1>[LATEXERROR::doc\engineering-reference\src\special-modules-reporting\resilience-metrics.tex:None::LaTeX Error: \begin{equation} on input line 1639 ended by \end{document}.]

These have been addressed in commit fc0bd265512b0fcd21a53adebb1536f9896ffeb2

sebastienbrideau committed:
Edits names of idd parameters

sebastienbrideau committed:
Edits names of idd parameters

sebastienbrideau committed:
Changes the way BIPVT collector dimensions (width and length) are entered.
Instead of user entering data as inputs, dimensions of parent surface are used. Parent surface must be rectangular.

sebastienbrideau committed:
Commits new test file with rectangular surfaces for BIPVT roof, to accommodate change in commit c85bc0105bbe9da675950f197498f7d4a7320acc

mjwitte commented:
@khaddad @sebastienbrideau Please fix the build warnings (most are unused variables) and bring the branch up to date. Also, please edit the top comment to include a brief description of this new feature.

amirroth commented:
Also please address my comments.

khaddad commented:
Thanks @mjwitte and @amirroth for all your comments. We'll do our best to address the remaining comments this week.

khaddad committed:
Define variable thisArray using auto option

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -103,11 +107,12 @@ namespace PhotovoltaicThermalCollectors {
+    int NumPVT(0); // count of all types of PVT in input file
+
     PlantComponent *PVTCollectorStruct::factory(EnergyPlusData &state, std::string_view objectName)
+
```
I missed this entire file, so there will be another set of comments incoming. Sorry.

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -165,60 +168,96 @@ namespace PhotovoltaicThermalCollectors {
+        int NumNumbers;          // Number of Numbers for each GetObjectItem call
+        int IOStatus;            // Used in GetObjectItem
+        bool ErrorsFound(false); // Set to true if errors in input, fatal at end of routine
+        using ScheduleManager::GetScheduleIndex;
```
Not using `using` statements anymore.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -165,60 +168,96 @@ namespace PhotovoltaicThermalCollectors {
+                                                                     state.dataIPShortCut->lAlphaFieldBlanks,
+                                                                     state.dataIPShortCut->cAlphaFieldNames,
+                                                                     state.dataIPShortCut->cNumericFieldNames);
+            if (UtilityRoutines::IsNameEmpty(state, state.dataIPShortCut->cAlphaArgs(1), state.dataIPShortCut->cCurrentModuleObject, ErrorsFound))
```
This check is not necessary. This gets caught by the epJSON parser, which prints an error and exits. If you got here, the name is not empty.

khaddad replied:
removed

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -165,60 +168,96 @@ namespace PhotovoltaicThermalCollectors {
+                                                                     state.dataIPShortCut->cNumericFieldNames);
+            if (UtilityRoutines::IsNameEmpty(state, state.dataIPShortCut->cAlphaArgs(1), state.dataIPShortCut->cCurrentModuleObject, ErrorsFound))
+                continue;
+            tmpSimplePVTperf(Item).Name = state.dataIPShortCut->cAlphaArgs(1);
```
Use `auto &thisTmpSimplePVTperf` shortcut.

khaddad replied:
updated

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -165,60 +168,96 @@ namespace PhotovoltaicThermalCollectors {
+            if (UtilityRoutines::IsNameEmpty(state, state.dataIPShortCut->cAlphaArgs(1), state.dataIPShortCut->cCurrentModuleObject, ErrorsFound))
+                continue;
+            tmpSimplePVTperf(Item).Name = state.dataIPShortCut->cAlphaArgs(1);
+            if (UtilityRoutines::SameString(state.dataIPShortCut->cAlphaArgs(2), "Fixed")) {
```
Use `getEnumerationValue` pattern. Many examples in the code.

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -165,60 +168,96 @@ namespace PhotovoltaicThermalCollectors {
+            tmpSimplePVTperf(Item).ThermEffic = state.dataIPShortCut->rNumericArgs(2);
+            tmpSimplePVTperf(Item).ThermEffSchedNum = ScheduleManager::GetScheduleIndex(state, state.dataIPShortCut->cAlphaArgs(3));
+            if ((tmpSimplePVTperf(Item).ThermEffSchedNum == 0) && (tmpSimplePVTperf(Item).ThermEfficMode == ThermEfficEnum::SCHEDULED)) {
+                ShowSevereError(state, "Invalid " + state.dataIPShortCut->cAlphaFieldNames(3) + " = " + state.dataIPShortCut->cAlphaArgs(3));
```
Use `format` instead of string concatenation. Many examples in the code.

khaddad replied:
updated to use format

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -233,7 +272,81 @@ namespace PhotovoltaicThermalCollectors {
                                                                      state.dataIPShortCut->cNumericFieldNames);
             if (UtilityRoutines::IsNameEmpty(state, state.dataIPShortCut->cAlphaArgs(1), state.dataIPShortCut->cCurrentModuleObject, ErrorsFound))
```
Same set of comments as previous function.

khaddad replied:
romoved

amirroth commented on src/EnergyPlus/PhotovoltaicThermalCollectors.cc:
```diff
@@ -233,7 +272,81 @@ namespace PhotovoltaicThermalCollectors {
+                                                                     state.dataIPShortCut->lAlphaFieldBlanks,
+                                                                     state.dataIPShortCut->cAlphaFieldNames,
+                                                                     state.dataIPShortCut->cNumericFieldNames);
+            if (UtilityRoutines::IsNameEmpty(state, state.dataIPShortCut->cAlphaArgs(1), state.dataIPShortCut->cCurrentModuleObject, ErrorsFound))
```
Same set of comments.

khaddad replied:
removed

khaddad committed:
Address comments from Mike W. and Amir R.

khaddad committed:
Address additional comments from Amir R.

khaddad committed:
Use 'DataGlobalConstants' for Pi

khaddad committed:
Merge remote-tracking branch 'upstream/develop' into bipvt

khaddad commented:
@mjwitte I just updated the branch and merged the latest E+ develop branch. Where exactly do I add a description of this new feature? Also I now don't see any warning locally after the build.

mjwitte commented:
> @mjwitte I just updated the branch and merged the latest E+ develop branch. Where exactly do I add a description of this new feature? 

At the very top of the conversations tab for https://github.com/NREL/EnergyPlus/pull/9594, in the initial comment under "Pull request overview". But maybe you don't have permission to edit comments? If so, just email me the description and I will add it.
![image](https://user-images.githubusercontent.com/4991075/212703972-6926ecdb-58b6-4cfe-ab40-35cb2eb2c09a.png)


> Also I now don't see any warning locally after the build.

It's quite common for the linux and Mac CI platforms to be more strict than windows. You can see the reports by following the "details" links in the "checks" summary at the bottom of the pull request page. Here are the latest reports from [Linux](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-c5e9e999a92a52ec4720027ea7e6273c8912456e-PR9594-x86_64-Linux-Ubuntu-18.04-gcc-7.5.html) and [Mac](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-c5e9e999a92a52ec4720027ea7e6273c8912456e-PR9594-x86_64-MacOS-10.17-clang-13.0.0.html).

khaddad committed:
Address warning and error messages from Linux builds

khaddad committed:
Apply clang format

khaddad commented:
@mjwitte I added a comment as requested. I see there is one more test pending but all the other ones have passed.

mjwitte commented:
> @mjwitte I added a comment as requested. I see there is one more test pending but all the other ones have passed.

Thanks. Linux still has one nit-picky warning.
![image](https://user-images.githubusercontent.com/4991075/212776090-5ff82cae-28a8-4d63-8040-04b6afb9fc80.png)

`for (i = 0; i <= (m ^ 2 - 1); i++) {`

Looks like the compiler is confused about the intent. Extra parentheses won't hurt, I suppose.

I'll do some more testing and re-read the docs, then we should be almost there.
And we need some unit tests for the new code.

khaddad commented:
Note we have this test case "ShopWithBIPVT.idf" in the "testfiles" folder. Do we need other test cases?

khaddad committed:
Address another warning from Libux build

mjwitte committed:
Initial unit tests for BIPVT

Commit:
Add unit tests for BIPVT model

khaddad committed:
Add enum class for PVT mode

khaddad committed:
Update PVT working fluid enum

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into bipvt

mjwitte committed:
clang format

mjwitte committed:
Make gcc happy

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into bipvt

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into bipvt

mjwitte commented on doc/input-output-reference/src/overview/group-solar-collectors.tex:
```diff
@@ -568,6 +568,117 @@ \subsubsection{Inputs}\label{inputs-5-029}
+
+This field is used to enter a nominal thickness of the air channel between the PV modules and the building. This is used to calculate the convective heat transfer coefficient behind the PV modules and on the building wall/roof surface. Units are \si{\m}. 
+
+\paragraph{Field: Effective Overall Height of Collector}\label{BIPVT-field-Height}
```
This field and next were removed from the IDD, so they need to be removed here. Also some other field names need to be updated to match the IDD. Also check the idf snippet below.

mjwitte commented on doc/input-output-reference/src/overview/group-solar-collectors.tex:
```diff
@@ -568,6 +568,117 @@ \subsubsection{Inputs}\label{inputs-5-029}
 
+\subsection{SolarCollectorPerformance:PhotovoltaicThermal:BIPVT}\label{solarcollectorperformancephotovoltaicthermalBIPVT}
+
+This object is used to provide performance details for the Building-Integrated PVT (BIPVT) model. There are no specific output variables for this object, reporting is done by the parent PVT object.
```
"parent PVT object". It would be helpful to spell out the full object type here with hyperref.

Myoldmopar replied:
Agreed, it would be nice to put a link in here.

mjwitte commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/solar-collectors.tex:
```diff
@@ -1274,3 +1274,436 @@ \subsubsection{References}\label{references-3-004}
 
 ISO. 2003. ISO 15099:2003. Thermal performance of windows, doors, and shading devices -- Detailed calculations. International Organization for Standardization.
+
+\subsection{Air-Based Building Integrated Parallel Flow PV/Thermal Collector}\label{building-integrated-pv-thermal-collector}
```
Thanks for the detailed engineering reference section!

mjwitte reviewed (commented):
@khaddad I've made some changes which will hopefully satisfy Linux gcc. There a few documentation changes needed. Otherwise I think this is in good shape.
@Myoldmopar Do you want to review code?

sebastienbrideau committed:
-adds new cases to previous unit tests.
-fixes thermal efficiency calculation bug in PhotovoltaicThermalCollectors.cc and adjust unit tests outputs

mjwitte committed:
format

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into bipvt

mjwitte committed:
Acknowledgment

Myoldmopar commented on src/EnergyPlus/Photovoltaics.cc:
```diff
@@ -246,6 +247,7 @@ namespace Photovoltaics {
         // Using/Aliasing
         using namespace DataHeatBalance;
 
+        using PhotovoltaicThermalCollectors::GetPVTmodelIndex;
```
We try to get rid of using statements.  It's not a red flag, but this is just one more that we have to remove later.

Myoldmopar commented on src/EnergyPlus/Photovoltaics.cc:
```diff
@@ -925,232 +940,220 @@ namespace Photovoltaics {
         Real64 Ee;
 
         ThisSurf = state.dataPhotovoltaic->PVarray(PVnum).SurfacePtr;
+        auto &thisPVarray = state.dataPhotovoltaic->PVarray(PVnum);
```
This is a very good use of a reference, thanks!

khaddad replied:
> OK, so this does look close. I think the biggest red flags for me are:
> 
> * what appears to be a global variable still living in the code
> * an Array1D being passed by value, so a copy is being made (I realize it's going to be a small array, but we still don't want that pattern in new code)
> 
> @khaddad I am grateful you have been so responsive, and I'd like to do whatever I can to help get this in for this release. Are you able to address the final comments here, or do I need to do it to get this moving?
> 
> Thanks for this feature!

@Myoldmopar Thanks for the offer to help move this contribution along. Please go ahead and make the suggested changes.

Myoldmopar reviewed (commented):
OK, so this does look close.  I think the biggest red flags for me are:
 - what appears to be a global variable still living in the code
 - an Array1D being passed by value, so a copy is being made (I realize it's going to be a small array, but we still don't want that pattern in new code)

@khaddad I am grateful you have been so responsive, and I'd like to do whatever I can to help get this in for this release.  Are you able to address the final comments here, or do I need to do it to get this moving?  

Thanks for this feature!

EnergyArchmage commented:
This looks like a great addition.  Model is a big improvement over simple PVT!

khaddad commented:
> This looks like a great addition. Model is a big improvement over simple PVT!

Thanks Brent for all the work you did on this part of E+

sebastienbrideau committed:
changes documentation to reflect changes to idd file.

sebastienbrideau committed:
Merge branch 'bipvt' of https://github.com/khaddad/EnergyPlus into bipvt

Myoldmopar committed:
Merge upstream develop, resolve conflict

Myoldmopar commented:
Alright, so the global variable that was left in there was only a dangling extern in the header that was never accessed.  I removed it, and it built fine.  I did a quick cleanup in the PV header to eliminate the member constructors and clean up some arguments.  I will be cleaning up the array copy argument issue as well.  I resolved the tiny conflict.  So far tests are looking good locally.  I'll be running final regressions and unit tests, and looking at the example file outputs and error file.  If everything is good, this can merge in right away.

Myoldmopar committed:
Cleanup in PV header and accompanying cc file

Myoldmopar committed:
Change args to c++ array references, apply formatting

Myoldmopar committed:
Merge branch 'bipvt' of github.com:khaddad/EnergyPlus into bipvt

Myoldmopar commented:
Everything is totally happy locally, not even any regressions -- those must have just appeared because of being outdated from develop.  Merging this.  Thanks @khaddad @sebastienbrideau and @mjwitte!
